src/buf_ring.rs

```diff
@@ -0,0 +1,87 @@
+use std::{convert::TryFrom, marker::PhantomData, num::Wrapping, slice, sync::atomic};
+
+use crate::{sys, util::Mmap};
+
+pub struct BufRing {
+    ring: Mmap,
+    ring_entries: u16,
+    entry_size: u32,
```

Comment:
you might want to decouple this in the future.
Although its terribly complicated, you might want to reuse parts of existing buffers.

Lets say you have 1k buffers, but a message was received with only 8 bytes. In theory you could immediately requeue the rest of that buffer onto the buf_ring for the next network data. The hard part is obviously accounting for this and soon you might have built an allocator internally - but possibly worth allowing for this use case in the future.

Reply:
Yeah, originally I wanted to enable full flexibility, but that made the API a pain to use for the simple case. I forgot to mention that for creation I decided against allowing you to specify arbitrary buffer sizes because that effectively limits you to the smallest buffer size since the kernel can't pick the buffer size it wants and has to go in order. Similar logic applies for what you described: if the data is tiny and processing will take a while, you should probably just copy it off the buffer, but if the data is large, then you'll be adding back a small fraction of the buffer to the ring making it have limited utility when the kernel tries to use it.

Anyway, that was the rationale. But I don't think this current architecture prevents flexibility in the future (or I could add it now). I think you could have an unsafe get method get returns a buffer struct and an unsafe add method that adds a buffer struct. Then we could special case entry_size of 0 to not initialize any of the buffers so you can allocate your own.

Reply:
i think that sounds generally sensible - was just in case you hadn't thought of it.

Reply:
Yeah, totally fair. :)